backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,163 @@
+// limitations under the License.
+
+package database
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+const DefaultViewRefreshIntervalSecond = 5
+
+type refreshFunc func(db *gorm.DB)
+
+type ViewRefreshHandler struct {
+	mu                        sync.Mutex
+	viewsToRefresh            map[string][]string // map of tables that shows which views should be refreshed due to table changes
+	tableChanged              map[string]bool
+	refreshFunc               map[string]refreshFunc // map of refresh functions for specified views
+	viewRefreshIntervalSecond time.Duration
+}
+
+func (vh *ViewRefreshHandler) TableChanged(table string) {
+	if !vh.IsSetViewRefreshHandler() {
+		return
+	}
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	vh.tableChanged[table] = true
+}
+
+func (vh *ViewRefreshHandler) GetAndClearChanges() map[string]bool {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	tables := vh.tableChanged
+	vh.tableChanged = make(map[string]bool)
+	return tables
+}
+
+func (vh *ViewRefreshHandler) runRequiredRefreshes(db *gorm.DB) {
+	viewNames := vh.getViewsToRefresh()
+	for viewName := range viewNames {
+		vh.refreshFunc[viewName](db)
+	}
+}
+
+// getViewsToRefresh creates a list of views that should be refreshed due to table changes.
+// The viewsToRefresh is the map of tables that shows which views should be refreshed due to table changes.
+func (vh *ViewRefreshHandler) getViewsToRefresh() map[string]bool {
+	viewToRefresh := make(map[string]bool)
+	tables := vh.GetAndClearChanges()
+	for table := range tables {
+		for _, viewName := range vh.viewsToRefresh[table] {
+			viewToRefresh[viewName] = true
+		}
+	}
+
+	return viewToRefresh
+}
+
+func (vh *ViewRefreshHandler) RegisterViewRefreshHandlers(dbDriver string) {
+	switch dbDriver {
+	case DBDriverTypePostgres:
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(applicationViewName),
+			applicationViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(resourceViewName),
+			resourceViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(packageViewName),
+			packageViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+		vh.registerViewRefreshHandler(getPostgresViewRefreshHandlerFunc(vulnerabilityViewName),
+			vulnerabilityViewName,
+			[]string{
+				applicationTableName,
+				resourceTableName,
+				packageTableName,
+				vulnerabilityTableName,
+			})
+	}
+}
+
+func (vh *ViewRefreshHandler) registerViewRefreshHandler(f refreshFunc, viewName string, tables []string) {
+	vh.refreshFunc[viewName] = f
+	vh.addViewsToRefreshByTable(viewName, tables)
```

Comment:
not sure why this needed to be a second function I think this could just be inlined.

I like having two maps one for the callbacks and then one to map the callbacks by name to the tables as it will allow us to keep track of all the callbacks which have been registered.

However I think we should change the viewsToRefresh map to a `map[string]<set>` to ensure that a callback can only be added once to a particular table the current append @L126 will not check this.

changing viewsToRefresh too:

```
type viewNameSet map[string]struct{}

type ViewRefreshHandler struct {
    ...
    viewsToRefresh map[string]viewNameSet
    ...
}
```

and then:

```
func (vh *ViewRefreshHandler) addViewsToRefreshByTable(viewName string, tables []string) {
	for _, table := range tables {
                if _, ok := vh.viewsToRefresh[table]; ok {
                        vh.viewsToRefresh[table] = viewNameSet{}
                }
	        vh.viewsToRefresh[table][viewName] = struct{}{}
	}
}
```

will ensure a view can only be in a table's list to refresh once and its safe to call the register function multiple times. To make it even safer to re-register we could even clear all other tables using that view like:

```
...
        for _, views := range vh.viewsToRefresh {
                delete(views, viewName)
        }
	for _, table := range tables {
                if _, ok := vh.viewsToRefresh[table]; ok {
                        vh.viewsToRefresh[table] = viewNameSet{}
                }
	        vh.viewsToRefresh[table][viewName] = struct{}{}
	}
...
```